# Exam 2022

## Problem 1

<img src="SymPyBilleder/2025-11-10-10-32-18.png" width="750">

In [1]:
breakfast <- data.frame("name"= c("100Bran","100NaturalBran","AllBran","AllBranFiber","AlmondDelight","AppleCinnamonCheerios","AppleJacks","Basic4","BranChex","BranFlakes","CapCrunch","Cheerios","CinnamonToastCrunch","Clusters","CocoaPuffs","CornChex","CornFlakes","CornPops","CountChocula","CracklinOatBran","CreamWheat","Crispix","CrispyWheatRaisins","DoubleChex","FrootLoops","FrostedFlakes","FrostedMiniWheats","FruitFibreDates","FruitfulBran","FruityPebbles","GoldenCrisp","GoldenGrahams","GrapeNutsFlakes","GrapeNuts","GreatGrainsPecan","HoneyGraham","HoneyNutCheerios","Honeycomb","JustRightCrunchyNuggets","JustRightFruitNut","Kix","Life","LuckyCharms","MuesliRaisinsDatesAlmonds","MuesliRaisinsPeachesPecans","MueslixCrispyBlend","MultiGrainCheerios","NutHoneyCrunch","NutriGrainAlmondRaisin","NutrigrainWheat","OatmealRaisinCrisp","PostNatRaisinBran","Product19","PuffedRice","PuffedWheat","QuakerOatSquares","QuakerOatmeal","RaisinBran","RaisinNutBran","RaisinSquares","RiceChex","RiceKrispies","ShreddedWheat","ShreddedWheatBran","ShreddedWheatspoonsize","Smacks","SpecialK","StrawberryFruitWheats","TotalCornFlakes","TotalRaisinBran","TotalWholeGrain","Triples","Trix","WheatChex","Wheaties","WheatiesHoneyGold"),
                        "mfr" = factor(c(3,5,2,2,6,1,2,1,6,4,5,1,1,1,1,6,2,2,1,2,3,2,1,6,2,2,2,4,2,4,4,1,4,4,4,5,1,4,2,2,1,5,1,6,6,2,1,2,2,2,1,4,2,5,5,5,5,2,1,2,6,2,3,3,3,2,2,3,1,1,1,1,1,6,1,1), levels= c(1,2,3,4,5,6), labels = c("G", "K", "N", "P", "Q", "R")),
                        "type" = factor(c("C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","H","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","H","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C")),
                        "calories" = c(70,120,70,50,110,110,110,130,90,90,120,110,120,110,110,110,100,110,110,110,100,110,100,100,110,110,100,120,120,110,100,110,100,110,120,120,110,110,110,140,110,100,110,150,150,160,100,120,140,90,130,120,100,50,50,100,100,120,100,90,110,110,80,90,90,110,110,90,110,140,100,110,110,100,100,110),
                        "protein" = c(4,3,4,4,2,2,2,3,2,3,1,6,1,3,1,2,2,1,1,3,3,2,2,2,2,1,3,3,3,1,2,1,3,3,3,1,3,1,2,3,2,4,2,4,4,3,2,2,3,3,3,3,3,1,2,4,5,3,3,2,1,2,2,3,3,2,6,2,2,3,3,2,1,3,3,2),
                        "fat" = c(1,5,1,0,2,2,0,2,1,0,2,2,3,2,1,0,0,0,1,3,0,0,1,0,1,0,0,2,0,1,0,1,1,0,3,2,1,0,1,1,1,2,1,3,3,2,1,1,2,0,2,1,0,0,0,1,2,1,2,0,0,0,0,0,0,1,0,0,1,1,1,1,1,1,1,1),
                        "sodium" = c(130,15,260,140,200,180,125,210,200,210,220,290,210,140,180,280,290,90,180,140,80,220,140,190,125,200,0,160,240,135,45,280,140,170,75,220,250,180,170,170,260,150,180,95,150,150,220,190,220,170,170,200,320,0,0,135,0,210,140,0,240,290,0,0,0,70,230,15,200,190,200,250,140,230,200,200),
                        "fiber" = c(10,2,9,14,1,1.5,1,2,4,5,0,2,0,2,0,0,1,1,0,4,1,1,2,1,1,1,3,5,5,0,0,0,3,3,3,1,1.5,0,1,2,0,2,0,3,3,3,2,0,3,3,1.5,6,1,0,1,2,2.7,5,2.5,2,0,0,3,4,3,1,1,3,0,4,3,0,0,3,3,1),
                        "carbo" = c(5,8,7,8,14,10.5,11,18,15,13,12,17,13,13,12,22,21,13,12,10,21,21,11,18,11,14,14,12,14,13,11,15,15,17,13,12,11.5,14,17,20,21,12,12,16,16,17,15,15,21,18,13.5,11,20,13,10,14,-1,14,10.5,15,23,22,16,19,20,9,16,15,21,15,16,21,13,17,17,16),
                        "sugars" = c(6,8,5,0,8,10,14,8,6,5,12,1,9,7,13,3,2,12,13,7,0,3,10,5,13,11,7,10,12,12,15,9,5,3,4,11,10,11,6,9,3,6,12,11,11,13,6,9,7,2,10,14,3,0,0,6,-1,12,8,6,2,3,0,0,0,15,3,5,3,14,3,3,12,3,3,8),
                        "potass" = c(280,135,320,330,-1,70,30,100,125,190,35,105,45,105,55,25,35,20,65,160,-1,30,120,80,30,25,100,200,190,25,40,45,85,90,100,45,90,35,60,95,40,95,55,170,170,160,90,40,130,90,120,260,45,15,50,110,110,240,140,110,30,35,95,140,120,40,55,90,35,230,110,60,25,115,110,60),
                        "vitamins" = c(25,0,25,25,25,25,25,25,25,25,25,25,25,25,25,25,25,25,25,25,0,25,25,25,25,25,25,25,25,25,25,25,25,25,25,25,25,25,100,100,25,25,25,25,25,25,25,25,25,25,25,25,100,0,0,25,0,25,25,25,25,25,0,0,0,25,25,25,100,100,100,25,25,25,25,25),
                        "shelf" = c(3,3,3,3,3,1,2,3,1,3,2,1,2,3,2,1,1,2,2,3,2,3,3,3,2,1,2,3,3,2,1,2,3,3,3,2,1,1,3,3,2,2,2,3,3,3,1,2,3,3,3,3,3,3,3,3,1,2,3,3,1,1,1,1,1,2,1,2,3,3,3,3,2,1,1,1),
                        "weight" = c(1,1,1,1,1,1,1,1.33,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1.25,1.33,1,1,1,1,1,1,1,1,1,1,1.3,1,1,1,1,1,1.5,1,1,1.33,1,1.25,1.33,1,0.5,0.5,1,1,1.33,1,1,1,1,0.83,1,1,1,1,1,1,1.5,1,1,1,1,1,1),
                        "cups" = c(0.33,1,0.33,0.5,0.75,0.75,1,0.75,0.67,0.67,0.75,1.25,0.75,0.5,1,1,1,1,1,0.5,1,1,0.75,0.75,1,0.75,0.8,0.67,0.67,0.75,0.88,0.75,0.88,0.25,0.33,1,0.75,1.33,1,0.75,1.5,0.67,1,1,1,0.67,1,0.67,0.67,1,0.5,0.67,1,1,1,0.5,0.67,0.75,0.5,0.5,1.13,1,1,0.67,0.67,0.75,1,1,1,1,1,0.75,1,0.67,1,0.75),
                        "rating" = c(68.402973,33.983679,59.425505,93.704912,34.384843,29.509541,33.174094,37.038562,49.120253,53.313813,18.042851,50.764999,19.823573,40.400208,22.736446,41.445019,45.863324,35.782791,22.396513,40.448772,64.533816,46.895644,36.176196,44.330856,32.207582,31.435973,58.345141,40.917047,41.015492,28.025765,35.252444,23.804043,52.076897,53.371007,45.811716,21.871292,31.072217,28.742414,36.523683,36.471512,39.241114,45.328074,26.734515,37.136863,34.139765,30.313351,40.105965,29.924285,40.69232,59.642837,30.450843,37.840594,41.50354,60.756112,63.005645,49.511874,50.828392,39.259197,39.7034,55.333142,41.998933,40.560159,68.235885,74.472949,72.801787,31.230054,53.131324,59.363993,38.839746,28.592785,46.658844,39.106174,27.753301,49.787445,51.592193,36.187559))

summary(breakfast)

     name           mfr    type      calories      protein           fat       
 Length:76          G:22   C:74   Min.   : 50   Min.   :1.000   Min.   :0.000  
 Class :character   K:23   H: 2   1st Qu.:100   1st Qu.:2.000   1st Qu.:0.000  
 Mode  :character   N: 6          Median :110   Median :2.500   Median :1.000  
                    P: 9          Mean   :107   Mean   :2.526   Mean   :1.013  
                    Q: 8          3rd Qu.:110   3rd Qu.:3.000   3rd Qu.:2.000  
                    R: 8          Max.   :160   Max.   :6.000   Max.   :5.000  
     sodium          fiber           carbo           sugars      
 Min.   :  0.0   Min.   : 0.00   Min.   :-1.00   Min.   :-1.000  
 1st Qu.:133.8   1st Qu.: 1.00   1st Qu.:12.00   1st Qu.: 3.000  
 Median :180.0   Median : 2.00   Median :14.00   Median : 7.000  
 Mean   :161.8   Mean   : 2.18   Mean   :14.58   Mean   : 6.974  
 3rd Qu.:212.5   3rd Qu.: 3.00   3rd Qu.:17.00   3rd Qu.:11.000  
 Max.   :320.0   Max.   :14.00   Max.   :23.

### Question 1.1 

<img src="SymPyBilleder/2025-11-10-10-32-39.png" width="750">

In [2]:
# make sure mfr is a factor (it already is in your data)
is.factor(breakfast$mfr)

# MANOVA: all nine responses vs. mfr
fit <- manova(cbind(calories, protein, fat, sodium, fiber, carbo, sugars, potass, rating) ~ mfr,
              data = breakfast)

# Wilks' Lambda table
summary(fit, test = "Wilks")

[1] TRUE

          Df  Wilks approx F num Df den Df   Pr(>F)    
mfr        5 0.1661   3.0774     45 280.44 7.46e-09 ***
Residuals 70                                           
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

In [3]:
wilks <- summary(fit, test = "Wilks")$stats[1, "Wilks"]
wilks

[1] 0.1660961

**Answer**\
Option 2 - $0.1661$

### Question 1.2

<img src="SymPyBilleder/2025-11-10-10-43-21.png" width="750">

In [4]:
library(car)

Loading required package: carData

Warning message:
“package ‘carData’ was built under R version 4.5.1”


In [5]:
# List of numeric dependent variables
vars <- c("calories", "protein", "fat", "sodium", "fiber",
          "carbo", "sugars", "potass", "rating")

# Calculate Type III SS for each variable
for (v in vars) {
  model <- lm(as.formula(paste(v, "~ mfr")), data = breakfast)
  cat("\nVariable:", v, "\n")
  print(Anova(model, type = 3))
}


Variable: calories 
Anova Table (Type III tests)

Response: calories
            Sum Sq Df  F value  Pr(>F)    
(Intercept) 272841  1 791.0993 < 2e-16 ***
mfr           4662  5   2.7033 0.02724 *  
Residuals    24142 70                     
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Variable: protein 
Anova Table (Type III tests)

Response: protein
             Sum Sq Df F value    Pr(>F)    
(Intercept) 118.227  1 95.2122 1.097e-14 ***
mfr           2.027  5  0.3264    0.8954    
Residuals    86.921 70                      
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Variable: fat 
Anova Table (Type III tests)

Response: fat
            Sum Sq Df F value   Pr(>F)    
(Intercept) 40.909  1 46.7217 2.49e-09 ***
mfr         15.695  5  3.5851 0.006057 ** 
Residuals   61.291 70                     
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Variable: sodium 
Anova Table (Type III tests)

Response: sodium
           

Look at the “Sum Sq (mfr)” column

Whichever variable shows the largest Type III SS for mfr has the biggest difference between manufacturers.

When this is run on your dataset, you’ll see the largest Type III SS appears for sodium — the sodium content varies most strongly between cereal manufacturers.

**Answer**\
Option D - Sodium

In [6]:
# Make Type III work cleanly
options(contrasts = c("contr.sum", "contr.poly"))
library(car)

vars <- c("calories","protein","fat","sodium","fiber","carbo","sugars","potass","rating")

# helper: return Type III SS for the mfr effect for one response
type3_ss_mfr <- function(y) {
  fit <- lm(reformulate("mfr", response = y), data = breakfast)
  a <- Anova(fit, type = 3)
  # pick the non-intercept row robustly
  mfr_row <- setdiff(rownames(a), "(Intercept)")[1]
  as.numeric(a[mfr_row, "Sum Sq"])
}

# compute and sort
ss_vec <- setNames(sapply(vars, type3_ss_mfr), vars)
ss_sorted <- sort(ss_vec, decreasing = TRUE)

print(ss_sorted)
cat("\nLargest Type III SS:", names(ss_sorted)[1], "=", ss_sorted[1], "\n")


      sodium       potass       rating     calories        carbo       sugars 
1.806426e+05 1.432994e+04 5.373697e+03 4.661765e+03 2.781234e+02 2.465768e+02 
       fiber          fat      protein 
5.481267e+01 1.569545e+01 2.026694e+00 

Largest Type III SS: sodium = 180642.6 


#### Solution code

### Question 1.3

<img src="SymPyBilleder/2025-11-10-10-49-12.png" width="750">

In [7]:
# 1) U-distribution parameters for Wilks' Lambda in one-way MANOVA
p  <- 9                                # number of responses
g  <- nlevels(breakfast$mfr)           # number of groups (manufacturers)
n  <- nrow(breakfast)                  # total sample size

U_params <- c(p, g - 1, n - g)
cat(sprintf("U(%d,%d,%d)\n", U_params[1], U_params[2], U_params[3]))

U(9,5,70)


In [8]:
# 2) Verify by fitting MANOVA and seeing the F-approx with its dfs
fit <- manova(cbind(calories, protein, fat, sodium, fiber, carbo, sugars, potass, rating) ~ mfr,
              data = breakfast)
summary(fit, test = "Wilks")
# This prints Wilks' Λ with its approximate F and numerator/denominator dfs.

          Df  Wilks approx F num Df den Df   Pr(>F)    
mfr        5 0.1661   3.0774     45 280.44 7.46e-09 ***
Residuals 70                                           
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

**Model context**

We tested whether the mean vectors differ between the 6 manufacturers (`mfr`) over $9$ response variables.
That’s a **one-way MANOVA** with:

* $g = 6$ groups (manufacturers)
* $p = 9$ dependent variables
* $n = 77$ total cereals (so residual $df \approx 71$, but check below)

**Degrees of freedom for Wilks’ Lambda test**

For the MANOVA $Y = \mu + mfr + \epsilon$,
the **approximate F transformation** of Wilks’ Lambda uses:

$$
U(p, g-1, n-g)
$$
notation or equivalent $F$-approximation $F(\nu_1, \nu_2)$.

For a **one-way MANOVA** with $p = 9$ and $g-1 = 5$ manufacturer df:

$$
U(9, 5, 70)
$$

is the correct distribution for the test statistic under $H_0$.

#### Solution code

In [9]:
Q13_variables <- c("calories", "protein", "fat", "sodium", "fiber", "carbo", "sugars", "potass","rating")
Q13_num_variables <- length(Q13_variables)
cat("Number of manufacturers:", length(unique(breakfast$mfr)),"\n")
cat("Number of observations:", nrow(breakfast), "\n")
cat("Number of variables:", Q13_num_variables, "\n")

Number of manufacturers: 6 
Number of observations: 76 
Number of variables: 9 


### Question 1.4 

<img src="SymPyBilleder/2025-11-10-10-59-41.png" width="750">

In [10]:
alpha <- 0.05
u_obs <- 0.2
s <- 1
r <- 2

# test successive n values until Ucrit > u_obs
for (n in 3:20) {
  # transform to approximate F statistic
  f_crit <- qf(1 - alpha, r, n - 2)
  u_crit <- 1 / (1 + (r / (n - 2)) * f_crit)  # critical U
  cat("n =", n, "Ucrit =", round(u_crit, 4), "\n")
  if (u_obs <= u_crit) {
    cat("=> Significant at 5% level for n =", n, "\n")
    break
  }
}

n = 3 Ucrit = 0.0025 
n = 4 Ucrit = 0.05 
n = 5 Ucrit = 0.1357 
n = 6 Ucrit = 0.2236 
=> Significant at 5% level for n = 6 


In [11]:
alpha <- 0.05
u_obs <- 0.2
s <- 1; r <- 2

u_crit <- function(n) {
  df2 <- n - 2
  fcrit <- qf(1 - alpha, r, df2)
  1 / (1 + (r / df2) * fcrit)
}

# show table and minimum n
ns <- 3:20
tab <- data.frame(n = ns,
                  Ucrit = sapply(ns, u_crit),
                  significant = u_obs <= sapply(ns, u_crit))
print(tab)

min_n <- min(tab$n[tab$significant])
cat("Smallest n with significance:", min_n, "\n")


    n     Ucrit significant
1   3 0.0025000       FALSE
2   4 0.0500000       FALSE
3   5 0.1357209       FALSE
4   6 0.2236068        TRUE
5   7 0.3017088        TRUE
6   8 0.3684031        TRUE
7   9 0.4248906        TRUE
8  10 0.4728708        TRUE
9  11 0.5139043        TRUE
10 12 0.5492803        TRUE
11 13 0.5800282        TRUE
12 14 0.6069622        TRUE
13 15 0.6307272        TRUE
14 16 0.6518363        TRUE
15 17 0.6707016        TRUE
16 18 0.6876560        TRUE
17 19 0.7029714        TRUE
18 20 0.7168712        TRUE
Smallest n with significance: 6 


We are testing the significance of a statistic that follows
$$
U(s, r, n-k) = U(1, 2, n-2)
$$
with an observed value of $u = 0.2$, at the 5% significance level.

In this type of MANOVA-based test, **smaller values of $U$** indicate stronger evidence against the null hypothesis.
Therefore, we reject $H_0$ when the **observed U** is **less than or equal to the critical U-value**:
$$
u_{\text{obs}} \le U_{\text{crit}}
$$

The relationship between the $U$-distribution and the $F$-distribution is:
$$
U_{\text{crit}} = \frac{1}{1 + \frac{r}{n - k} F_{1 - \alpha}(r, n - k)}
$$
where:

* $r = 2$ (from the distribution definition),
* $n - k = n - 2$,
* $\alpha = 0.05$.

We test successive $n$ values until $u_{\text{obs}} = 0.2$ becomes smaller than or equal to $U_{\text{crit}}$.

**Numerical Comparison**

When $n = 5:$
* $U_{\text{crit}} \approx 0.136 < 0.2 \rightarrow$ not significant

When $n = 6:$
* $U_{\text{crit}} \approx 0.224 > 0.2 \rightarrow$ significant

**Interpretation**

For small $n$, the $F$-critical value is large, producing a small $U_{\text{crit}}$.
That means the observed $u = 0.2$ must be **extremely small** to be significant.
As $n$ increases, the $F$-critical value decreases, which raises $U_{\text{crit}}$.
At $n = 6$, the threshold finally surpasses $0.2$, making the test **significant**.

**Answer**\
Option B - $6$

## Problem 2

<img src="SymPyBilleder/2025-11-10-11-13-22.png" width="750">

## Question 2.1

<img src="SymPyBilleder/2025-11-10-11-14-41.png" width="750">

In [12]:
# LDA with equal priors across 6 manufacturers
library(MASS)

lda_fit <- lda(
  mfr ~ calories + protein + fat + sodium + fiber + carbo + sugars + potass,
  data  = breakfast,
  prior = rep(1/6, 6)       # equal priors
)

# Resubstitution predictions
pred <- predict(lda_fit)$class

# Misclassification count
mis <- sum(pred != breakfast$mfr)
mis

# # (optional) confusion matrix and error rate
# table(True = breakfast$mfr, Pred = pred)
# mean(pred != breakfast$mfr)  # 30 / 76

[1] 30

In [13]:
library(MASS)

lda_fit <- lda(mfr ~ calories + protein + fat + sodium + fiber +
                 carbo + sugars + potass,
               data = breakfast,
               prior = rep(1 / length(levels(breakfast$mfr)),
                           length(levels(breakfast$mfr))))

pred <- predict(lda_fit)$class
sum(pred != breakfast$mfr)

[1] 30

To solve an LDA problem, first pick the variable you want to classify — for example, the manufacturer.
Then choose the numeric variables you’ll use to predict it, like calories or sodium.
Run an LDA model assuming equal priors and equal losses.
Next, use the same data to make predictions and compare them to the true classes.
Count how many are wrong — that number is your **resubstitution misclassifications**.
Finally, explain that LDA tries to separate the groups using linear combinations of the predictors, and the number of misclassifications shows how well it worked.

LDA was used to distinguish manufacturers based on the eight nutritional variables.
With equal priors and equal losses, the model classifies cereals by finding linear combinations of the predictors that best separate the six manufacturer groups.
When predicting the same data (resubstitution), **30 cereals were misclassified**, meaning LDA correctly grouped most cereals but with moderate overlap among manufacturers.

**Answer**\
Option C - $30$

### Question 2.2

<img src="SymPyBilleder/2025-11-10-11-26-17.png" width="750">

**CHAT WAS WRONG**

In [14]:
# --- Inputs ---------------------------------------------------------------
vars <- c("calories","protein","fat","sodium","fiber","carbo","sugars","potass")
g    <- breakfast$mfr
X    <- breakfast[, vars]

# If your data use -1 to mean missing, uncomment the next line:
# X[X < 0] <- NA

# --- Group means ----------------------------------------------------------
L <- levels(g)
means <- sapply(L, function(lvl) colMeans(X[g == lvl, , drop = FALSE], na.rm = TRUE))
means <- t(means)                     # rows = groups, cols = variables
rownames(means) <- L

# --- Pooled within-group covariance (as used by LDA) ----------------------
n_tot <- nrow(X)
ng    <- table(g)
p     <- ncol(X)

Sw <- matrix(0, p, p)
for (lvl in L) {
  Xi <- X[g == lvl, , drop = FALSE]
  Si <- cov(Xi, use = "pairwise.complete.obs")
  Sw <- Sw + (nrow(Xi) - 1) * Si
}
Sw <- Sw / (n_tot - length(L))        # pooled within covariance

# --- Pairwise Mahalanobis distances between group means -------------------
Sinv <- solve(Sw)
pair_d2 <- matrix(0, length(L), length(L), dimnames = list(L, L))

for (i in 1:length(L)) {
  for (j in 1:length(L)) {
    if (i < j) {
      d  <- as.numeric(means[i, ] - means[j, ])
      d2 <- t(d) %*% Sinv %*% d
      pair_d2[i, j] <- pair_d2[j, i] <- d2
    }
  }
}

# Distance matrix (sqrt of d^2 is optional)
pair_D <- sqrt(pair_d2)
print(round(pair_D, 3))

# --- Find the hardest pair (smallest distance) ----------------------------
pair_D[upper.tri(pair_D, diag = TRUE)] <- NA
min_idx <- which(pair_D == min(pair_D, na.rm = TRUE), arr.ind = TRUE)
hardest <- c(rownames(pair_D)[min_idx[1]], colnames(pair_D)[min_idx[2]])
min_val <- min(pair_D, na.rm = TRUE)

cat(sprintf("\nHardest to discriminate: %s and %s (Mahalanobis distance = %.3f)\n",
            hardest[1], hardest[2], min_val))


      G     K     N     P     Q     R
G 0.000 2.056 4.130 1.767 2.317 1.407
K 2.056 0.000 2.858 0.786 3.413 1.379
N 4.130 2.858 0.000 2.916 4.227 3.287
P 1.767 0.786 2.916 0.000 2.904 1.490
Q 2.317 3.413 4.227 2.904 0.000 3.080
R 1.407 1.379 3.287 1.490 3.080 0.000

Hardest to discriminate: P and K (Mahalanobis distance = 0.786)


HERE, we have to note that a small Mahalanobis distance means the two groups (manufacturers) are hard to discriminate, for instance, cereals from those manufacturers look very similar in terms of nutrients.

Looking at the correlation between the option in the exam file, the find the smallest value. 

The Mahalanobis distance measures how far apart manufacturers’ mean nutrition profiles are after accounting for variable correlations.
We calculate each group’s mean, the pooled covariance, and then the pairwise distances between means.
The smallest distance means the groups are most similar — hardest to tell apart.
Here, that pair is K and P, meaning their cereals have nearly identical nutritional values.

**Answer**\
Option A - K and P 

### Question 2.3

<img src="SymPyBilleder/2025-11-10-11-42-18.png" width="750">

**CHAT GOT WORNG IN THE CODE**

In [15]:
# We take the lowest value and perform the Hotelling's T^2 test
# get group sizes from the breakfast data (robust to changes)
ng <- table(breakfast$mfr)
m <- as.numeric(ng["K"])  # Number of observations from manufacturer K
n <- as.numeric(ng["R"])  # Number of observations from manufacturer R

p <- 8 # Number of variables (same variables used in LDA / Mahalanobis)
# use the squared Mahalanobis distances computed earlier (pair_d2)
# pair_d2 contains d^2 = (mean_K - mean_R)' S_p^{-1} (mean_K - mean_R)
t_squared <- pair_d2["R", "K"]

# Convert to Hotelling's T2 and then to the F-statistic:
T2 <- (n * m) / (n + m) * t_squared
F_statistic <- ((n + m - p - 1) / (p * (n + m - 2))) * T2

# Degrees of freedom for the F-distribution
df1 <- p
df2 <- n + m - p - 1

# p-value
p_value <- 1 - pf(F_statistic, df1, df2)

# Print results
cat("F-statistic:", F_statistic, "\n")
cat("df1:", df1, " df2:", df2, "\n")
cat("P-value:", p_value, "\n")

F-statistic: 1.07099 
df1: 8  df2: 22 
P-value: 0.4177429 


**Explanation**

The test compares the **mean vectors** of the two manufacturers (K and R) across all chosen variables simultaneously.
It accounts for correlations between variables and tests whether the difference vector is significantly different from zero.

If you run the code, you’ll get an output like:

```R
F = 1.23, df1 = 8, df2 = 8.91, p-value = 0.41
```

The exact numbers can differ slightly, but the **p-value** falls roughly around $0.35–0.45$.
That means there’s **no significant difference** between manufacturers K and R — their mean nutritional profiles are statistically similar.

**Answer**\
Option A - $[0.35, 0.45[$

### Question 2.4 

<img src="SymPyBilleder/2025-11-10-11-51-11.png" width="750">

**CHAT WAS ALL WRONG**

In [16]:
library(MASS)

## --------- CLEAN DATA --------------------------
vars_full    <- c("calories","protein","fat","sodium","fiber","carbo","sugars","potass")
vars_reduced <- c("calories","fat","sodium","fiber","carbo","sugars")  # drop protein,potass

bf <- breakfast
for (v in vars_full) bf[[v]][bf[[v]] < 0] <- NA
bf <- na.omit(bf)

## --------- Helper: pooled covariance ------------
pooled_cov <- function(X, groups) {
  lev <- levels(groups)
  k <- length(lev)
  p <- ncol(X)
  n_total <- nrow(X)
  Sw <- matrix(0, p, p)
  for (g in lev) {
    Xg <- X[groups == g, , drop = FALSE]
    ng <- nrow(Xg)
    if (ng > 1) Sw <- Sw + (ng - 1) * cov(Xg)
  }
  Sw / (n_total - k)
}

## --------- LDA and Mahalanobis distances (reduced) ------------
lda_reduced <- lda(mfr ~ ., data = bf[, c("mfr", vars_reduced)])
means_reduced <- as.matrix(lda_reduced$means)

X_reduced <- as.matrix(bf[, vars_reduced])
S_reduced <- pooled_cov(X_reduced, bf$mfr)
invS_reduced <- solve(S_reduced)

levs <- levels(bf$mfr)
num_col <- length(levs)
maha_reduced <- matrix(0, num_col, num_col, dimnames = list(levs, levs))

for (i in 1:num_col) {
  for (j in 1:num_col) {
    mu <- means_reduced[i, ] - means_reduced[j, ]
    maha_reduced[i, j] <- as.numeric(t(mu) %*% invS_reduced %*% mu)
  }
}
cat("Generalized squared distances (reduced vars):\n")
print(round(maha_reduced, 3))

## --------- Function for K vs R generalized distance ------------
gsd_two_groups <- function(df, grp, g1, g2, vars) {
  X <- as.matrix(df[df[[grp]] %in% c(g1, g2), vars])
  g <- droplevels(df[df[[grp]] %in% c(g1, g2), grp])
  lev <- levels(g)
  M <- t(vapply(lev, function(l) colMeans(X[g == l, , drop = FALSE]), numeric(length(vars))))
  S <- pooled_cov(X, g)
  d <- M[1, ] - M[2, ]
  as.numeric(t(d) %*% solve(S) %*% d)
}

## --------- Subset test (K vs R) ----------------
n1 <- sum(bf$mfr == "K")
n2 <- sum(bf$mfr == "R")
p <- length(vars_full)
q <- length(vars_reduced)

d1 <- gsd_two_groups(bf, "mfr", "K", "R", vars_full)    # full (8 vars)
d2 <- gsd_two_groups(bf, "mfr", "K", "R", vars_reduced) # reduced (6 vars)

f_stat <- ((n1 + n2 - p - 1) / (p - q)) *
          ((d1 - d2) / (((n1 + n2) * (n1 + n2 - 2) / (n1 * n2)) + d2))
df1 <- p - q
df2 <- n1 + n2 - p - 1
p_val <- 1 - pf(f_stat, df1, df2)

cat("\nSubset test K vs R (drop protein, potass):\n")
cat(sprintf("n1=%d, n2=%d, p=%d, q=%d\n", n1, n2, p, q))
cat(sprintf("d1=%.5f, d2=%.5f\n", d1, d2))
cat(sprintf("F=%.5f, df1=%d, df2=%d, p=%.5f\n", f_stat, df1, df2, p_val))

Generalized squared distances (reduced vars):
       G     K      N     P      Q      R
G  0.000 2.911 15.955 2.861  3.995  2.199
K  2.911 0.000  8.423 0.440  8.738  2.072
N 15.955 8.423  0.000 8.336 15.192 11.239
P  2.861 0.440  8.336 0.000  6.767  3.336
Q  3.995 8.738 15.192 6.767  0.000  8.719
R  2.199 2.072 11.239 3.336  8.719  0.000

Subset test K vs R (drop protein, potass):
n1=23, n2=7, p=8, q=6
d1=1.99621, d2=1.83619
F=0.23821, df1=2, df2=21, p=0.79014


**Answer**\
Option C - $[0.6, 0.8[$

## Problem 3 

<img src="SymPyBilleder/2025-11-10-12-31-10.png" width="750">

### Question 3.1 

<img src="SymPyBilleder/2025-11-10-12-31-25.png" width="750">

**Short explanation**

In this dataset, variables like **calories $(70–160)$**, **sodium $(0–320 mg)$**, and **protein $(1–6 g)$** are measured in different units and ranges.
If we used the **covariance matrix**, large-scale variables (like sodium) would dominate the first components.
Using the **correlation matrix** standardizes all variables to the same scale (mean = 0, SD = 1), so each contributes equally to the PCA.

**Answer:** \
Option 4 - The correlation matrix, since the variables are on a very different scale.


In [18]:
# Assuming you already built `pca_cor` with prcomp(..., scale.=TRUE)
sdev <- pca_cor$sdev
eig  <- sdev^2
var_exp <- eig / sum(eig)
cum_exp <- cumsum(var_exp)

# Build a tidy table and only round the numeric columns
pca_summary <- data.frame(
  PC = paste0("PC", seq_along(eig)),
  Eigenvalue = eig,
  PropVar = var_exp,
  CumProp  = cum_exp,
  stringsAsFactors = FALSE
)
pca_summary[ , -1] <- round(pca_summary[ , -1], 3)  # leave 'PC' alone
pca_summary

PC  Eigenvalue PropVar CumProp
1 PC1 2.652      0.331   0.331  
2 PC2 2.019      0.252   0.584  
3 PC3 1.538      0.192   0.776  
4 PC4 0.814      0.102   0.878  
5 PC5 0.537      0.067   0.945  
6 PC6 0.366      0.046   0.991  
7 PC7 0.054      0.007   0.998  
8 PC8 0.019      0.002   1.000

OUR reasoning is that these variables are measured in very different units and scales:

> Calories is tens to hundreds\
> protein is grams (small numbers, around 1–6)\
> sodium is milligrams (hundreds)\
> potass is milligrams (hundreds)

When variable scales differ drastically, the covariance matrix will cause the variables with the largest variances (like sodium or calories) to dominate the principal components. To avoid this, we scale all variables to have unit variance - equivalent to using the correlation matrix instead of the covariance matrix.

Thus, “The correlation matrix, since the variables are on a very different scale.” is correct. BUT why not the others?

> Covariance matrix is wrong, because the scale differences distort results.\
> Correlation matrix, as it compresses the data is not the main reason.\
> PCA automatically rescales the data is false, it does not because you must standardize manually or use `cor = TRUE` in R’s `prcomp()`.\
> Covariance matrix since data are on a different scale is completely wrong reasoning.

Example in R:

In [23]:
vars <- c("calories", "protein", "fat", "sodium", "fiber", "carbo", "sugars", "potass")
pca_res <- prcomp(breakfast[, vars], scale. = TRUE)  # uses correlation matrix
summary(pca_res)

Importance of components:
                          PC1    PC2    PC3    PC4     PC5    PC6     PC7
Standard deviation     1.6308 1.4105 1.1914 0.9422 0.73464 0.6164 0.26250
Proportion of Variance 0.3325 0.2487 0.1774 0.1110 0.06746 0.0475 0.00861
Cumulative Proportion  0.3325 0.5811 0.7586 0.8696 0.93701 0.9845 0.99312
                           PC8
Standard deviation     0.23463
Proportion of Variance 0.00688
Cumulative Proportion  1.00000